internal/attestation/gcp/snp/issuer.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package snp
+
+import (
+	"context"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+
+	"github.com/edgelesssys/constellation/v2/internal/attestation"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/gcp"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/snp"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/variant"
+	"github.com/edgelesssys/constellation/v2/internal/attestation/vtpm"
+
+	"github.com/google/go-sev-guest/abi"
+	sevclient "github.com/google/go-sev-guest/client"
+	"github.com/google/go-tpm-tools/client"
+	tpmclient "github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/proto/attest"
+)
+
+// Issuer issues SEV-SNP attestations.
+type Issuer struct {
+	variant.GCPSEVSNP
+	*vtpm.Issuer
+}
+
+// NewIssuer creates a SEV-SNP based issuer for GCP.
+func NewIssuer(log attestation.Logger) *Issuer {
+	return &Issuer{
+		Issuer: vtpm.NewIssuer(
+			vtpm.OpenVTPM,
+			getAttestationKey,
+			getInstanceInfo,
+			log,
+		),
+	}
+}
+
+// getAttestationKey returns a new attestation key.
+func getAttestationKey(tpm io.ReadWriter) (*tpmclient.Key, error) {
+	tpmAk, err := client.GceAttestationKeyRSA(tpm)
+	if err != nil {
+		return nil, fmt.Errorf("creating RSA Endorsement key: %w", err)
+	}
+
+	return tpmAk, nil
+}
+
+// getInstanceInfo generates an extended SNP report, i.e. the report and any loaded certificates.
+// Report generation is triggered by sending ioctl syscalls to the SNP guest device, the AMD PSP generates the report.
+// The returned bytes will be written into the attestation document.
+func getInstanceInfo(_ context.Context, tpm io.ReadWriteCloser, _ []byte) ([]byte, error) {
+	tpmAk, err := client.GceAttestationKeyRSA(tpm)
+	if err != nil {
+		return nil, fmt.Errorf("creating RSA Endorsement key: %w", err)
+	}
+
+	encoded, err := x509.MarshalPKIXPublicKey(tpmAk.PublicKey())
+	if err != nil {
+		return nil, fmt.Errorf("marshalling public key: %w", err)
+	}
+
+	akDigest := sha512.Sum512(encoded)
+
+	device, err := sevclient.OpenDevice()
+	if err != nil {
+		return nil, fmt.Errorf("opening sev device: %w", err)
+	}
+	defer device.Close()
+
+	report, certs, err := sevclient.GetRawExtendedReportAtVmpl(device, akDigest, 0)
```

Comment:
```suggestion
func getInstanceInfo(_ context.Context, tpm io.ReadWriteCloser, extraData []byte) ([]byte, error) {
	device, err := sevclient.OpenDevice()
	if err != nil {
		return nil, fmt.Errorf("opening sev device: %w", err)
	}
	defer device.Close()

	report, certs, err := sevclient.GetRawExtendedReportAtVmpl(device, extraData, 0)
```
We should generate the report over `extraData` (issuer TLS public key hash + nonce)